File: apps/ios/retailmanagement/Models/ManagerContracts.swift
import Foundation

// MARK: - Enums

enum RecommendationType: String, Codable, Sendable {
    case reorder
    case priceChange = "price_change"
    case stockAnomaly = "stock_anomaly"
}

enum RecommendationStatus: String, Codable, Sendable {
    case pending
    case approved
    case dismissed
}

enum InventoryType: String, Codable, Sendable {
    case finished
    case material
    case purchased
}

enum SourcingStrategy: String, Codable, Sendable {
    case supplierPremade = "supplier_premade"
    case manufacturedStandard = "manufactured_standard"
    case manufacturedCustom = "manufactured_custom"
}

// Support for arbitrary metrics mapping in Recommendation
enum MetricValue: Codable, Sendable {
    case string(String)
    case number(Double)
    case boolean(Bool)
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let x = try? container.decode(String.self) {
            self = .string(x)
            return
        }
        if let x = try? container.decode(Double.self) {
            self = .number(x)
            return
        }
        if let x = try? container.decode(Bool.self) {
            self = .boolean(x)
            return
        }
        throw DecodingError.typeMismatch(MetricValue.self, DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Wrong type for MetricValue"))
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let x): try container.encode(x)
        case .number(let x): try container.encode(x)
        case .boolean(let x): try container.encode(x)
        }
    }
}

// MARK: - Manager Summary

struct RecommendationOutcome: Codable, Sendable, Identifiable {
    var id: String { recommendationId }
    
    let recommendationId: String
    let skuId: String
    let title: String
    let type: RecommendationType
    let status: RecommendationStatus
    let updatedAt: String
}

struct ManagerSummary: Codable, Sendable {
    let storeId: String
    let analysisStatus: String
    let lastGeneratedAt: String
    let lowStockCount: Int
    let anomalyCount: Int
    let pendingPriceRecommendations: Int
    let pendingReorderRecommendations: Int
    let pendingStockAnomalies: Int
    let openPurchaseOrders: Int
    let activeWorkOrders: Int
    let inTransitTransfers: Int
    let purchasedUnits: Int
    let materialUnits: Int
    let finishedUnits: Int
    let recentOutcomes: [RecommendationOutcome]
}

struct SupplyChainSummary: Codable, Sendable {
    let storeId: String
    let supplierCount: Int
    let openPurchaseOrders: Int
    let activeWorkOrders: Int
    let inTransitTransfers: Int
    let purchasedUnits: Int
    let materialUnits: Int
    let finishedUnits: Int
    let openRecommendationLinkedOrders: Int?
}

// MARK: - Detailed Insights

struct InventoryInsight: Codable, Sendable, Identifiable {
    var id: String { skuId }
    
    let inventoryId: String?
    let skuId: String
    let storeId: String
    let skuCode: String
    let description: String
    let longDescription: String?
    let inventoryType: InventoryType
    let sourcingStrategy: SourcingStrategy
    let supplierName: String?
    let costPrice: Double?
    let currentPrice: Double?
    let currentPriceValidUntil: String?
    let purchasedQty: Int
    let purchasedIncomingQty: Int
    let materialQty: Int
    let materialIncomingQty: Int
    let materialAllocatedQty: Int
    let finishedQty: Int
    let finishedAllocatedQty: Int
    let inTransitQty: Int
    let activeWorkOrderCount: Int
    let qtyOnHand: Int
    let reorderLevel: Int
    let reorderQty: Int
    let lowStock: Bool
    let anomalyFlag: Bool
    let anomalyReason: String?
    let recentSalesQty: Int
    let recentSalesRevenue: Double
    let avgDailySales: Double
    let daysOfCover: Double?
    let pendingRecommendationCount: Int
    let pendingPriceRecommendationCount: Int
    let lastUpdated: String
}

struct ManagerRecommendation: Codable, Sendable, Identifiable {
    let id: String
    let storeId: String
    let skuId: String
    let inventoryId: String?
    let inventoryType: InventoryType
    let sourcingStrategy: SourcingStrategy
    let supplierName: String?
    let type: RecommendationType
    let status: RecommendationStatus
    let title: String
    let rationale: String
    let confidence: Double
    let supportingMetrics: [String: MetricValue]
    let source: String
    let expectedImpact: String
    let currentPrice: Double?
    let suggestedPrice: Double?
    let suggestedOrderQty: Int?
    let workflowAction: String
    let analysisStatus: String
    let generatedAt: String
    let decidedAt: String?
    let appliedAt: String?
    let note: String?
}

// MARK: - Tracking Entities

struct InventoryAdjustmentHistory: Codable, Sendable, Identifiable {
    let id: String
    let inventoryId: String
    let skuId: String
    let storeId: String
    let quantityDelta: Int
    let resultingQty: Int
    let reason: String
    let source: String
    let note: String?
    let createdAt: String
}

struct SupplierSummary: Codable, Sendable, Identifiable {
    let id: String
    let name: String
    let contactName: String?
    let email: String?
    let phone: String?
    let leadTimeDays: Int?
    let currency: String?
    let notes: String?
    let isActive: Bool
}

struct StageInventoryPosition: Codable, Sendable, Identifiable {
    let id: String
    let storeId: String
    let skuId: String
    let skuCode: String
    let description: String
    let inventoryType: InventoryType
    let sourcingStrategy: SourcingStrategy
    let supplierName: String?
    let quantityOnHand: Int
    let incomingQuantity: Int
    let allocatedQuantity: Int
    let availableQuantity: Int
}

struct PurchaseOrderLine: Codable, Sendable, Identifiable {
    var id: String { lineId }
    let lineId: String
    let skuId: String
    let skuCode: String
    let description: String
    let stageInventoryType: InventoryType
    let quantity: Int
    let unitCost: Double
    let receivedQuantity: Int
    let openQuantity: Int
    let note: String?
}

struct PurchaseOrderSummary: Codable, Sendable, Identifiable {
    let id: String
    let supplierId: String
    let supplierName: String
    let status: String
    let lines: [PurchaseOrderLine]
    let totalQuantity: Int
    let totalCost: Double
    let expectedDeliveryDate: String?
    let note: String?
    let recommendationId: String?
}

struct BOMRecipeComponent: Codable, Sendable {
    let skuId: String
    let skuCode: String
    let description: String
    let quantityRequired: Double
    let note: String?
}

struct BOMRecipeSummary: Codable, Sendable, Identifiable {
    let id: String
    let storeId: String
    let finishedSkuId: String
    let finishedSkuCode: String
    let finishedDescription: String
    let name: String
    let yieldQuantity: Int
    let components: [BOMRecipeComponent]
    let notes: String?
}

struct WorkOrderComponent: Codable, Sendable {
    let skuId: String
    let skuCode: String
    let description: String
    let quantityRequired: Double
    let note: String?
}

struct WorkOrderSummary: Codable, Sendable, Identifiable {
    let id: String
    let finishedSkuId: String
    let finishedSkuCode: String
    let finishedDescription: String
    let workOrderType: String
    let status: String
    let targetQuantity: Int
    let completedQuantity: Int
    let components: [WorkOrderComponent]
    let dueDate: String?
    let note: String?
    let recommendationId: String?
}

struct StockTransferSummary: Codable, Sendable, Identifiable {
    let id: String
    let skuId: String
    let skuCode: String
    let description: String
    let quantity: Int
    let fromInventoryType: InventoryType
    let toInventoryType: InventoryType
    let status: String
    let note: String?
    let recommendationId: String?
    let dispatchedAt: String?
    let receivedAt: String?
}

File: apps/staff-portal/src/lib/manager-contracts.ts
export type RecommendationType = "reorder" | "price_change" | "stock_anomaly";

export type RecommendationStatus =
  | "pending"
  | "approved"
  | "rejected"
  | "applied"
  | "expired"
  | "queued"
  | "unavailable";

export type InventoryType = "purchased" | "material" | "finished";

export type SourcingStrategy =
  | "supplier_premade"
  | "manufactured_standard"
  | "manufactured_custom";

export interface RecommendationOutcome {
  recommendation_id: string;
  sku_id: string | null;
  title: string;
  type: RecommendationType;
  status: RecommendationStatus;
  updated_at: string | null;
}

export interface ManagerSummary {
  store_id: string;
  analysis_status: string;
  last_generated_at: string | null;
  low_stock_count: number;
  anomaly_count: number;
  pending_price_recommendations: number;
  pending_reorder_recommendations: number;
  pending_stock_anomalies: number;
  open_purchase_orders: number;
  active_work_orders: number;
  in_transit_transfers: number;
  purchased_units: number;
  material_units: number;
  finished_units: number;
  recent_outcomes: RecommendationOutcome[];
}

export interface InventoryInsight {
  inventory_id: string | null;
  sku_id: string;
  store_id: string;
  sku_code: string;
  description: string;
  long_description: string | null;
  inventory_type: InventoryType;
  sourcing_strategy: SourcingStrategy;
  supplier_name: string | null;
  cost_price: number | null;
  current_price: number | null;
  current_price_valid_until: string | null;
  purchased_qty: number;
  purchased_incoming_qty: number;
  material_qty: number;
  material_incoming_qty: number;
  material_allocated_qty: number;
  finished_qty: number;
  finished_allocated_qty: number;
  in_transit_qty: number;
  active_work_order_count: number;
  qty_on_hand: number;
  reorder_level: number;
  reorder_qty: number;
  low_stock: boolean;
  anomaly_flag: boolean;
  anomaly_reason: string | null;
  recent_sales_qty: number;
  recent_sales_revenue: number;
  avg_daily_sales: number;
  days_of_cover: number | null;
  pending_recommendation_count: number;
  pending_price_recommendation_count: number;
  last_updated: string | null;
}

export interface ManagerRecommendation {
  id: string;
  store_id: string;
  sku_id: string | null;
  inventory_id: string | null;
  inventory_type: InventoryType;
  sourcing_strategy: SourcingStrategy;
  supplier_name: string | null;
  type: RecommendationType;
  status: RecommendationStatus;
  title: string;
  rationale: string;
  confidence: number;
  supporting_metrics: Record<string, unknown>;
  source: string;
  expected_impact: string | null;
  current_price: number | null;
  suggested_price: number | null;
  suggested_order_qty: number | null;
  workflow_action: string | null;
  analysis_status: string;
  generated_at: string;
  decided_at: string | null;
  applied_at: string | null;
  note: string | null;
}

export interface InventoryAdjustmentHistory {
  id: string;
  inventory_id: string;
  sku_id: string;
  store_id: string;
  quantity_delta: number;
  resulting_qty: number;
  reason: string;
  source: string;
  note: string | null;
  created_at: string;
}

export interface SupplyChainSummary {
  store_id: string;
  supplier_count: number;
  open_purchase_orders: number;
  active_work_orders: number;
  in_transit_transfers: number;
  purchased_units: number;
  material_units: number;
  finished_units: number;
  open_recommendation_linked_orders: number;
}

export interface Supplier {
  id: string;
  name: string;
  contact_name: string | null;
  email: string | null;
  phone: string | null;
  lead_time_days: number;
  currency: string;
  notes: string | null;
  is_active: boolean;
}

export interface StageInventoryPosition {
  id: string;
  store_id: string;
  sku_id: string;
  sku_code: string;
  description: string;
  inventory_type: InventoryType;
  sourcing_strategy: SourcingStrategy;
  supplier_name: string | null;
  quantity_on_hand: number;
  incoming_quantity: number;
  allocated_quantity: number;
  available_quantity: number;
}

export interface PurchaseOrderLine {
  line_id: string;
  sku_id: string;
  sku_code: string;
  description: string;
  stage_inventory_type: InventoryType;
  quantity: number;
  unit_cost: number;
  received_quantity: number;
  open_quantity: number;
  note: string | null;
}

export interface PurchaseOrder {
  id: string;
  supplier_id: string;
  supplier_name: string | null;
  status: "draft" | "ordered" | "partially_received" | "received" | "cancelled";
  lines: PurchaseOrderLine[];
  total_quantity: number;
  total_cost: number;
  expected_delivery_date: string | null;
  note: string | null;
  recommendation_id: string | null;
}

export interface WorkOrderComponent {
  sku_id: string;
  sku_code: string;
  description: string;
  quantity_required: number;
  note: string | null;
}

export interface WorkOrder {
  id: string;
  finished_sku_id: string;
  finished_sku_code: string;
  finished_description: string;
  work_order_type: "standard" | "custom";
  status: "draft" | "scheduled" | "in_progress" | "completed" | "cancelled";
  target_quantity: number;
  completed_quantity: number;
  components: WorkOrderComponent[];
  due_date: string | null;
  note: string | null;
  recommendation_id: string | null;
}

export interface BOMRecipeComponent {
  sku_id: string;
  sku_code: string;
  description: string;
  quantity_required: number;
  note: string | null;
}

export interface BOMRecipe {
  id: string;
  store_id: string;
  finished_sku_id: string;
  finished_sku_code: string;
  finished_description: string;
  name: string;
  yield_quantity: number;
  components: BOMRecipeComponent[];
  notes: string | null;
}

export interface StockTransfer {
  id: string;
  sku_id: string;
  sku_code: string;
  description: string;
  quantity: number;
  from_inventory_type: InventoryType;
  to_inventory_type: InventoryType;
  status: "draft" | "in_transit" | "received" | "cancelled";
  note: string | null;
  recommendation_id: string | null;
  dispatched_at: string | null;
  received_at: string | null;
}

export type DataEnvelope<T> = {
  data: T;
};
